File: cpp/Day3.cpp
#include "Day3.h"
#include "Utils.h"

#include <cstdio>
#include <string>

Day3::Day3()
{
    printf ("Day 3\n");

    std::filesystem::path contentPath = "/Users/reubenduckering/Documents/Personal Repo/Advent-Of-Code-2024/files/day_3.txt";
    const auto lines = Utils::getContentLines (contentPath);

    processInput (lines, validInstructions);

    printf("");
}

void Day3::runPart1()
{
    int sum = 0;

    for (const auto& instruction :  validInstructions)
    {
        if (instruction.starts_with ("mul("))
        {
            const size_t startIndex = instruction.find ('(');
            const size_t seperatorIndex = instruction.find (',');
            const size_t endIndex = instruction.find (')');

            const std::string firstNumber = instruction.substr (startIndex + 1, seperatorIndex - startIndex - 1);
            const std::string endNumber   = instruction.substr (seperatorIndex + 1, endIndex - seperatorIndex - 1);

            sum += std::stoi (firstNumber) * std::stoi (endNumber);
        }
    }

    printf("Sum: %i\n", sum);
}

void Day3::runPart2()
{
    int sum = 0;

    bool sumArguments = true;

    for (const auto& instruction :  validInstructions)
    {

        if (instruction.starts_with ("don't"))
        {
            sumArguments = false;
        }
        else if (instruction.starts_with ("do"))
        {
            sumArguments = true;
        }

        if (! sumArguments)
            continue;

        if (instruction.starts_with ("mul("))
        {
            const size_t startIndex = instruction.find ('(');
            const size_t seperatorIndex = instruction.find (',');
            const size_t endIndex = instruction.find (')');

            const std::string firstNumber = instruction.substr (startIndex + 1, seperatorIndex - startIndex - 1);
            const std::string endNumber   = instruction.substr (seperatorIndex + 1, endIndex - seperatorIndex - 1);

            sum += std::stoi (firstNumber) * std::stoi (endNumber);
        }
    }

    printf("Sum: %i\n", sum);
}

bool isWord (const std::array<char, 12>& buffer, const char* word)
{
    if (std::memcmp (&buffer, word, sizeof (char) * strlen (word)) == 0)
        return true;

    return false;
}

void Day3::processInput (const std::vector<std::string>& contentLines, std::vector<std::string>& instructions)
{
    std::array<char, 12> scratchBuffer;

    for (const auto& line : contentLines)
    {
        int valid_character_count = 0;

        for (const auto& c : line)
        {
            if (c == 'm')
            {
                valid_character_count = 0;
                scratchBuffer[valid_character_count] = c;
                continue;
            }

            if (c == 'd')
            {
                valid_character_count = 0;
                scratchBuffer[valid_character_count] = c;
                continue;
            }

            if (valid_character_count >= scratchBuffer.size() - 1)
                continue;

            scratchBuffer[++valid_character_count] = c;

            // end of sequence
            if (c == ')')
            {
                if (isWord (scratchBuffer, "don't"))
                {
                    instructions.emplace_back ("don't");
                }
                else if (isWord (scratchBuffer, "do"))
                {
                    instructions.emplace_back ("do");
                }
                else if (isWord (scratchBuffer, "mul("))
                {
                    char a[valid_character_count + 1];

                    std::memcpy (&a, &scratchBuffer, sizeof(char) * valid_character_count + 1);
                    const std::string valueToCheck = std::string (a);

                    const size_t startIndex = valueToCheck.find ('(');
                    const size_t commaIndex = valueToCheck.find (',');
                    const size_t endIndex = valueToCheck.find (')');

                    if (startIndex > endIndex)
                        continue;

                    if (commaIndex == -1)
                        continue;

                    if (commaIndex - startIndex > 4 || commaIndex - startIndex <= 1)
                        continue;

                    if (endIndex - commaIndex > 4 || endIndex - commaIndex <= 1)
                        continue;

                    instructions.emplace_back (valueToCheck);
                    valid_character_count = 0;
                }
            }
        }
    }
}

File: Python/src/day_3.py
from src import utils
import numpy as np

# solution is avoiding using regex!

class Day3:
    def __init__(self):
        self.file_contents = utils.read_file ("../files/day_3.txt")
        self.instructions = self.parse_content (self.file_contents)

    @staticmethod
    def parse_content(contents):
        lines = contents.splitlines()
        
        instructions = []

        for l in lines:
            for i in range(0, len(l), 1):
                buffer = l[i: i + 12]

                if buffer[:4] == "mul(":
                    print ("maybe")

            # for i in l.split(")"):

            #     if "mul" in i:
            #         mul_index = i.find ("mul(")
            #         instructions.append(i[mul_index:] + ')')

            #     elif "don't" in i:
            #         dont_index = i.find ("don't")
            #         instructions.append(i[dont_index:] + ')')

            #     elif "do" in i:
            #         do_index = i.find ("do")
            #         instructions.append(i[do_index:] + ')')

        return instructions
    
    def run_part_1(self):
        sum = 0
        for i in self.instructions:
            if "mul(" in i:
                start_index = i.find("(")
                comma_index = i.find(",")
                end_index = i.find(")")

                first_number = i[start_index + 1: comma_index]
                second_number = i[comma_index + 1: end_index]

                if len(first_number) > 3 or len(first_number) < 1:
                    continue;
                
                if len(second_number) > 3 or len(second_number) < 1:
                    continue;

                sum += int(first_number) * int(second_number)
        
        print(f"Sum: {sum}")

    def run_part_2(self):
        pass
        